docs/rfcs/xxxx-backwards-compatibility.md

```diff
@@ -0,0 +1,111 @@
+# Grafana Agent backwards compatibility 
+
+* Date: 2023-05-25
+* Author: Robert Fratto (@rfratto)
+* PR: [grafana/agent#3981](https://github.com/grafana/agent/pull/3981)
+* Status: Draft
+
+Grafana Agent has been following [semantic versioning](https://semver.org/) since its inception.
+After three years of development and 33 minor releases, the project is on trajectory to have a 1.0 release. 
+
+In the context of semantic versioning, a 1.0 release indicates that future minor releases have backwards compatibility with older minor releases in the same major release; version 1.1 is backwards compatible with version 1.0. Having major and minor releases signals to users when an upgrade may take more time (major releases) and when they can upgrade with more confidence (minor releases).
+
+However, Grafana Agent is a large project, with a large surface area for what may be considered part of the backwards compatibility guarantees. This proposal formally establishes what parts of Grafana Agent will be protected by backwards compatibility. 
+
+## Goals 
+
+- Set expectations for what is covered by backwards compatibility. 
+- Set expectations for when upgrades to new major versions will be forced.
+
+## Proposal 
+
+Backwards compatibility means that a user can upgrade their version of Grafana Agent without needing to make any changes with the way they interact with Grafana Agent, provided that interaction is within scope of being covered by backwards compatibility.
+
+### Scope of backwards compatibility   
+
+The following will be protected by backwards compatibility between minor releases: 
+
+- **User configuration**, including configuration file syntax and the command-line interface. 
+
+- **Versioned network APIs**, if any versioned APIs are introduced prior to the 1.0 release.
+
+- **Telemetry data used in official dashboards**. This means that users will continue to be able to use the same set of dashboards we provide when upgrading minor releases.    
+
+  - Official dashboards are dashboards in the repository's `operations/` [directory](../../operations/).
+
+- **Externally importable Go packages**. If a user is importing our code as a dependency, they should be able to upgrade to a new minor release without having to make changes to their code.
```

Comment:
Instead of using Agent as a library how about instead moving the Agent systems that would make good libraries into their own repositories. Then our usage would be the exact same as people wanting to use the subsystems of the Agent as a library. Then we could semantically version the subsystems independently of each other while maintaining that the Agent itself does not expose a stable api. This would also promote more usable, composable systems instead of tight coupling.

Reply:
> My concern is that we will have API changes that are unavoidable which will then trigger major version changes at a rate we do not like.

If we expect we need to have API changes that are unavoidable, we should not be exposing those APIs.

Reply:
That's interesting. It seems like such a case isn't covered by [Go's](https://go.dev/doc/go1compat) backwards compatibility guarantees either: 

> Methods. As with struct fields, it may be necessary to add methods to types. Under some circumstances, such as when the type is embedded in a struct along with another type, the addition of the new method may break the struct by creating a conflict with an existing method of the other embedded type. We cannot protect against this rare case and do not guarantee compatibility should it arise. 

They don't specifically call out the case you lined out, but it's the same. 

I think we should link out to Go's compatibility guarantees for our API compatibility guarantees to say we're a superset of this.

Reply:
I've changed the language to explicitly say that our expectations around backwards compatible API changes match Go's.

Reply:
Sounds reasonable